libc/socket/unix-socket.c

```diff
@@ -1480,6 +1481,189 @@ TEST(test_unix_socket, flags)
+	const char *socket_name = "/tmp/wrong_port";
+
+
+	fd[0] = open(socket_name, O_CREAT | O_WRONLY | O_TRUNC, 0666);
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(0, fd[0]);
+
+	fd[1] = socket(AF_UNIX, type, 0);
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(0, fd[1]);
+
+	addr.sun_family = AF_UNIX;
+	snprintf(addr.sun_path, sizeof(addr.sun_path), "%s", socket_name);
+
+	errno = 0;
+	err = connect(fd[1], (struct sockaddr *)&addr, SUN_LEN(&addr));
+	TEST_ASSERT_EQUAL_INT(-1, err);
+	TEST_ASSERT_EQUAL_INT(ENOTSOCK, errno);
+
+	if (type == SOCK_DGRAM) {
+		errno = 0;
+		err = sendto(fd[1], "data", 4, 0, (struct sockaddr *)&addr, SUN_LEN(&addr));
+		TEST_ASSERT_EQUAL_INT(-1, err);
+		TEST_ASSERT_EQUAL_INT(ENOTSOCK, errno);
+	}
+
+	close(fd[0]);
+	close(fd[1]);
+	unlink(socket_name);
+}
+
+
+TEST(test_unix_socket, wrong_port)
+{
+	unix_wrong_port(SOCK_STREAM);
+	unix_wrong_port(SOCK_SEQPACKET);
+	unix_wrong_port(SOCK_DGRAM);
+}
+
+
+static void unix_wrong_type(int type)
+{
+	int fd[2], err, wrong_type;
+	struct sockaddr_un addr = { 0 };
+	const char *socket_name = "/tmp/wrong_type";
+
+	switch (type) {
+		case SOCK_STREAM:
+			wrong_type = SOCK_SEQPACKET;
+			break;
+		case SOCK_SEQPACKET:
+			wrong_type = SOCK_DGRAM;
+			break;
+		case SOCK_DGRAM:
+			wrong_type = SOCK_STREAM;
+			break;
+		default:
+			FAIL("invalid_type");
+	}
+
+	if ((fd[0] = unix_named_socket(wrong_type, socket_name)) < 0)
+		FAIL("unix_named_socket");
+
+	fd[1] = socket(AF_UNIX, type, 0);
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(0, fd[1]);
+
+	addr.sun_family = AF_UNIX;
+	snprintf(addr.sun_path, sizeof(addr.sun_path), "%s", socket_name);
+
+	errno = 0;
+	err = connect(fd[1], (struct sockaddr *)&addr, SUN_LEN(&addr));
+	TEST_ASSERT_EQUAL_INT(-1, err);
+	TEST_ASSERT_EQUAL_INT(EPROTOTYPE, errno);
+
+	if (type == SOCK_DGRAM) {
+		errno = 0;
+		err = sendto(fd[1], "data", 4, 0, (struct sockaddr *)&addr, SUN_LEN(&addr));
+		TEST_ASSERT_EQUAL_INT(-1, err);
+		TEST_ASSERT_EQUAL_INT(EPROTOTYPE, errno);
+	}
+
+	close(fd[0]);
+	close(fd[1]);
+	unlink(socket_name);
+}
+
+
+TEST(test_unix_socket, wrong_type)
+{
+	unix_wrong_type(SOCK_STREAM);
+	unix_wrong_type(SOCK_SEQPACKET);
+	unix_wrong_type(SOCK_DGRAM);
+}
+
+
+TEST(test_unix_socket, clear_peer_closed)
+{
+	int fd[2], err;
+	struct sockaddr_un addr = { 0 };
+	const char *socket_name = "/tmp/clear_peer_closed";
+
+	if ((fd[0] = unix_named_socket(SOCK_DGRAM, socket_name)) < 0)
+		FAIL("unix_named_socket");
+
+	fd[1] = socket(AF_UNIX, SOCK_DGRAM, 0);
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(0, fd[1]);
+
+	addr.sun_family = AF_UNIX;
+	snprintf(addr.sun_path, sizeof(addr.sun_path), "%s", socket_name);
+
+	errno = 0;
+	err = connect(fd[1], (struct sockaddr *)&addr, SUN_LEN(&addr));
+	TEST_ASSERT_EQUAL_INT(0, err);
+	TEST_ASSERT_EQUAL_INT(0, errno);
+
+	close(fd[0]);
+	unlink(socket_name);
+
+	errno = 0;
+	err = send(fd[1], "data", 4, 0);
+	TEST_ASSERT_EQUAL_INT(-1, err);
+	TEST_ASSERT_EQUAL_INT(ECONNREFUSED, errno);
+
+	errno = 0;
+	err = send(fd[1], "data", 4, 0);
```

Comment:
Maybe also check `recv()`, `sendmsg()` etc.?

Reply:
I’ll add a new test for `recv()`. `sendmsg()` just calls `send()` under the hood.

Reply:
> `sendmsg()` just calls `send()` under the hood.

Yes, but (unit) tests should not rely on the implementation details